### PR TITLE
New system for autogenerating spawn points

### PR DESCRIPTION
**Pull request by Yolopanther12**
Imeplemented a new system for automatically generating spawn points for
players on maps with not enough spawn points.
The code used comes from TTT because it has an implementation that works
very well. It has been tweaked to better suit Prophunters.
The basic flow is:
1. Try to find a spawnpoint for a player
2. If not enough room, try to generate a new point nearby

Prophunters previously had a system for this but it mostly just killed players or got them stuck in each other. This system works much better and has good success generating new spawnpoints. It is still possible for players to get stuck in walls or objects but that seemed to only happen in extreme circumstances.

### CODE REVIEW

File: gamemode/sv_player.lua [hunk 1]
```diff
@@ -203,81 +203,243 @@ function GM:PlayerDeathSound()
...
-
-	// recount
-	local Count = table.Count( spawnPoints )
-	
-	if ( Count == 0 ) then
-		Msg("[PlayerSelectSpawn] Error! No spawn points!\n")
-		return nil
+
+-- This is only a shallow copy.
+local function MergeTables(...)
```
Comment (zikaeroh) [inline gamemode/sv_player.lua]: This isn't properly followed in a couple of places in the existing code, but if it's `local`, then I'd prefer it to be `camelCase` to better distinguish that it's defined in a specific file. Similarly for other functions / variables being added, if possible.

There are a few exceptions like `PlayerMeta` or `EntityMeta` which are named for classes, though.

File: gamemode/sv_player.lua [hunk 2]
```diff
@@ -203,81 +203,243 @@ function GM:PlayerDeathSound()
...
+
+local hunter_spawn_types = {"info_player_counterterrorist",
+"info_player_allies", "info_player_rebel", "info_player_knight",
+"diprip_start_team_red", "info_player_red", "info_player_zombie"}
+
+-- These spawn types should ideally only be used for spectators. Hunters/props spawning here
+-- will probably fall out of the world.
+local spectator_spawn_types = {"info_player_start", "gmod_player_start",
+"info_player_teamspawn", "ins_spawnpoint", "aoc_spawnpoint",
+"dys_spawn_point", "info_player_coop", "info_player_deathmatch"}
```
Comment (zikaeroh) [inline gamemode/sv_player.lua]: Any reason these variables are `snake_case`?

File: gamemode/sv_player.lua [hunk 3]
```diff
@@ -203,81 +203,243 @@ function GM:PlayerDeathSound()
...
+
+					ErrorNoHalt("PROPHUNTERS WARNING: Map has too few spawn points, using a rigged spawn for ".. tostring(ply:Nick()) .. "\n")
+
+					self.HaveRiggedSpawn = true
+					return rig_spwn
+				end
+			end
+		end
+	end
+
```
Comment (zikaeroh) [inline gamemode/sv_player.lua:429]: At the very last final step where it's known that players will be killed, could you send a message to at least the admins to say "something's wrong"?

File: gamemode/sv_player.lua [hunk 4]
```diff
@@ -203,81 +203,243 @@ function GM:PlayerDeathSound()
...
+"diprip_start_team_red", "info_player_red", "info_player_zombie"}
+
+-- These spawn types should ideally only be used for spectators. Hunters/props spawning here
+-- will probably fall out of the world.
+local spectator_spawn_types = {"info_player_start", "gmod_player_start",
+"info_player_teamspawn", "ins_spawnpoint", "aoc_spawnpoint",
+"dys_spawn_point", "info_player_coop", "info_player_deathmatch"}
+
+
+function GetSpawnEnts(ply_team, shuffled, force_all)
```
Comment (zikaeroh) [inline gamemode/sv_player.lua]: This isn't your code, but "shuffled == true" implies to me that the something is already shuffled ("shuffled is true"). Perhaps `shuffle` to indicate that something should be shuffled.

Better yet, eliminate this parameter as it's only ever called with `true`.

File: gamemode/sv_player.lua [hunk 5]
```diff
@@ -203,81 +203,243 @@ function GM:PlayerDeathSound()
...
+	local tbl_to_use
```
Comment (zikaeroh) [inline gamemode/sv_player.lua]: This must all be from TTT, since it's using `snake_case`.

File: gamemode/sv_player.lua [hunk 6]
```diff
@@ -203,81 +203,243 @@ function GM:PlayerDeathSound()
...
+	local pos = rigged && spwn || spwn:GetPos()
+
+	if not util.IsInWorld(pos) then return false end
+
+	local blocking = ents.FindInBox(pos + Vector( -32, -32, 0 ), pos + Vector( 32, 32, 64 )) -- Changed from (-16, -16, 0) (16, 16, 64)
+
+	for _, blocking_ent in ipairs(blocking) do
+		if IsValid(blocking_ent) && blocking_ent:IsPlayer() && blocking_ent:Team() != 1 && blocking_ent:Alive() then
+			if force then
+				blocking_ent:Kill()
```
Comment (zikaeroh) [inline gamemode/sv_player.lua]: Is this the code that does the too-many-players killing? Candidate for the message?

File: gamemode/sv_player.lua (hunk 5 above)
Comment (Yolopanther12) [reply gamemode/sv_player.lua]: All of the snake_case was totally my fault, I completely forgot that it should've been camelCase.

File: gamemode/sv_player.lua (hunk 6 above)
Comment (Yolopanther12) [reply gamemode/sv_player.lua]: Yep, this is it. I'm going to look into how I can send a message to admins only.

File: gamemode/sv_player.lua (hunk 2 above)
Comment (Yolopanther12) [reply gamemode/sv_player.lua]: My mind completely blanking and forgetting it should be camelCase.

File: gamemode/sv_player.lua (hunk 3 above)
Comment (Yolopanther12) [reply gamemode/sv_player.lua:429]: Added a system for this.

File: gamemode/sv_player.lua [hunk 7]
```diff
@@ -203,81 +203,250 @@ function GM:PlayerDeathSound()
...
+
+	for _, blockingEnt in ipairs(blocking) do
+		if IsValid(blockingEnt) && blockingEnt:IsPlayer() && blockingEnt:Team() != 1 && blockingEnt:Alive() then
+			if force then
+				blockingEnt:Kill()
+
+				blockingEnt:PlayerChatMsg(Color(200, 20, 20), "You were killed because there are not enough spawnpoints.")
+				for _, value in ipairs(player.GetAll()) do
+					if value:IsAdmin() || value:IsSuperAdmin() then
+						value:PlayerChatMsg(Color(200, 20, 20), "Not enough spawnpoints; " .. blockingEnt:Nick() .. " has been killed to spawn " .. ply:Nick())
```
Comment (zikaeroh) [inline gamemode/sv_player.lua]: Mega nit: needs a `.` at the end of the message to match other strings.

File: gamemode/sv_player.lua (hunk 4 above)
Comment (Yolopanther12) [reply gamemode/sv_player.lua]: I've gotten rid of the shuffled argument, the final list is now just always shuffled.

**Pull request merged**